userspace/libsinsp/container_engine/docker_common.cpp

```diff
@@ -339,79 +337,25 @@ bool docker::resolve(sinsp_container_manager* manager, sinsp_threadinfo* tinfo,
 	}
 
 	tinfo->m_container_id = container_id;
-
-	existing_container_info = manager->get_container(container_id);
-
-	if(!existing_container_info)
-	{
-		// Add a minimal container_info object where only the
-		// container id, (possibly) name, and a container
-		// image = incomplete is filled in. This may be
-		// overidden later once parse_docker_async completes.
-		sinsp_container_info container_info;
-
-		g_logger.format(sinsp_logger::SEV_DEBUG,
-				"docker_async (%s): No existing container info, creating initial stub info",
-				container_id.c_str());
-
-		container_info.m_type = CT_DOCKER;
-		container_info.m_id = container_id;
-		container_info.m_name = container_name;
-		container_info.m_image = s_incomplete_info_name;
-		container_info.m_imageid = s_incomplete_info_name;
-		container_info.m_imagerepo = s_incomplete_info_name;
-		container_info.m_imagetag = s_incomplete_info_name;
-		container_info.m_imagedigest = s_incomplete_info_name;
-		container_info.m_metadata_complete = false;
-
-		manager->add_container(container_info, tinfo);
-
-		existing_container_info = manager->get_container(container_id);
-	}
+	container_info = manager->get_container(container_id);
 
 #ifdef HAS_CAPTURE
 	// Possibly start a lookup for this container info
-	if(!existing_container_info->m_metadata_complete &&
-	    query_os_for_missing_info)
+	if(container_info == nullptr || container_info->query_anyway(CT_DOCKER))
```

Comment:
I feel like this could have a better comment about intention. Why are we sometimes creating a incomplete container?

Reply:
Does the comment for `query_anyway` help? https://github.com/draios/sysdig/pull/1446/files#diff-80cbd69f90deb31413ffe9e5b15a3786R202

The point of incomplete metadata in container_manager is so that we know a lookup is already in progress so we don't need to start another one.

We still have the old race somewhat as we can only store info about one pending lookup (not one per engine) though the async framework should handle this without duplicating actual queries. Maybe we need to rethink the way to store the "already running" flag but I'd say we're good for now.